src/cats/CatSlideshow.tsx

```diff
@@ -0,0 +1,178 @@
+import { useEffect, useCallback, useState } from "react";
+import getCatsApiFetchParams from "./getCatsApiFetchParams";
+import usePaginatedFetch, {
+  FetchPageCallbackType,
+  PageType,
+} from "./usePaginatedFetch";
+import usePrevious from "./usePrevious";
+import usePrefetchImages from "./usePrefetchImages";
+import CatSlideshowControls from "./CatSlideshowControls";
+import { SlideAnimation, Direction } from "./SlideAnimation";
+import { LoadingCard } from "./Cards";
+import styles from "./CatSlideshow.module.css";
+
+type CatSlideshowPropsType = {
+  selectedBreedID: string;
+};
+
+type CatAPISearchResponseItemType = {
+  id: string;
+  url: string;
+};
+
+type CatAPISearchResponseType = Array<CatAPISearchResponseItemType>;
+type SlideshowPageType = Array<{ id: string; url: string }>;
+type SlideshowPagesType = Record<number, PageType<SlideshowPageType, string>>;
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH_LIMIT = 8;
+
+function fetchPageFromCatsApi(
+  pageIndex: number,
+  selectedBreedID: string,
+  fetchPage: FetchPageCallbackType<SlideshowPageType>
+) {
+  const { url, options } = getCatsApiFetchParams("/images/search", {
+    breed_id: selectedBreedID === "all" ? "" : selectedBreedID,
+    limit: String(PAGE_SIZE),
+    order: "ASC",
+    page: String(pageIndex),
+  });
+
+  fetchPage({
+    url,
+    options,
+    index: pageIndex,
+    key: selectedBreedID,
+    getPageData: async (response): Promise<SlideshowPageType> => {
+      const json: CatAPISearchResponseType = await response.json();
+      return json.map((item) => ({ id: item.id, url: item.url }));
+    },
+    getMetadata: (response): Record<string, string> => {
+      const count = response.headers.get("pagination-count");
+      return count !== null ? { paginationCount: count } : {};
+    },
+  });
+}
+
+function getPageIndex(index: number) {
+  return Math.floor(index / PAGE_SIZE);
+}
+
+function getImageURL(pages: SlideshowPagesType, index: number) {
+  const pageIndex = getPageIndex(index);
+  const offset = index % PAGE_SIZE;
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+export default function CatSlideshow({
+  selectedBreedID,
+}: CatSlideshowPropsType) {
+  const [direction, setDirection] = useState(Direction.Next);
+  const [index, setIndex] = useState<number>(0);
+  const [visibleIndex, setVisibleIndex] = useState<number | null>(null);
+  const [maxIndex, setMaxIndex] = useState<number | null>(null);
+  const previousBreedID = usePrevious<string>(selectedBreedID);
+  const { pages, metadata, status, fetchPage, resetPages } = usePaginatedFetch<
```

Comment:
So going back to the earlier comment, I think what I would do here is to restore the reducer in `CatSlideshow` and have an option for `usePaginatedFetch` to call callbacks. Then you can just wire the up directly to a dispatch on your CatSlideshow reducer instead of trying to update CatSlideshow state in response to pagination state changing

Reply:
I'm surprised you suggest passing callbacks into PaginatedFetch. Isn't the whole point to define state once and then write useEffects() to synchronize the propagation of state changes? At the simplest end, look at useLocalStateStorage. It defines its own state and returns the state to the caller. 

I do also see that employing useEffect is adding additional renders needlessly. The imperative callbacks just feel odd to me.

Reply:
On further reflection, I might as well just pass the dispatch pointer from CatSlideshow into usePaginatedFetch. It's stable and it functions the same way you would if you passed it into a subcomponent.